### PR TITLE
take care of the nullable to-many relation in serializer

### PR DESCRIPTION
**Pull request by fherbin**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

don't throw UnexpectedValueException('Unexpected non-iterable value for to-many relation.') if the entity property is nullable (and null).
-> for my application, if a to-many relation is set to null, it's because I didn't hydrate that sub-resource.  (i use custom repositories/dataProviders)

### CODE REVIEW

File: src/Serializer/AbstractItemNormalizer.php [hunk 1]
```diff
@@ -523,6 +523,9 @@ protected function getAttributeValue($object, $attribute, $format = null, array
             $this->resourceClassResolver->isResourceClass($className)
         ) {
             if (!is_iterable($attributeValue)) {
+                if ($type->isNullable() && null === $attributeValue) {
```
Comment (dunglas) [inline src/Serializer/AbstractItemNormalizer.php]:
```suggestion
                if (null === $attributeValue && $type->isNullable()) {
```

(the comparison costs less than the function call)

**Comment by fherbin** [comment]
> Could to also add a unit test please?

I'm a beginner when it comes to testing with phpunit. and testing the serializer may not be the easiest way to learn.

Can you help me?

This is the test I've created, but I can't make it throw the execption, i dont know why.

```php
    public function testNullableToManyRelationProperty()
    {
        $dummy = new DummyTableInheritanceRelated();

        $propertyNameCollectionFactoryProphecy = $this->prophesize(PropertyNameCollectionFactoryInterface::class);
        $propertyNameCollectionFactoryProphecy->create(DummyTableInheritanceRelated::class, [])->willReturn(
            new PropertyNameCollection(['children'])
        )->shouldBeCalled();

        $nullablePropertyMetadataFactoryProphecy = $this->prophesize(PropertyMetadataFactoryInterface::class);
        $nullablePropertyMetadataFactoryProphecy->create(DummyTableInheritanceRelated::class, 'children', [])->willReturn(
            new PropertyMetadata(new Type(Type::BUILTIN_TYPE_ARRAY, true), '', true, true, true, true, false, null, null, null)
        )->shouldBeCalled();

        $notNullablePropertyMetadataFactoryProphecy = $this->prophesize(PropertyMetadataFactoryInterface::class);
        $notNullablePropertyMetadataFactoryProphecy->create(DummyTableInheritanceRelated::class, 'children', [])->willReturn(
            new PropertyMetadata(new Type(Type::BUILTIN_TYPE_ARRAY, false), '', true, true, true, true, false, null, null, null)
        )->shouldBeCalled();

        $iriConverterProphecy = $this->prophesize(IriConverterInterface::class);
        $iriConverterProphecy->getIriFromItem($dummy)->willReturn('/dummies/1')->shouldBeCalled();

        $propertyAccessorProphecy = $this->prophesize(PropertyAccessorInterface::class);
        $propertyAccessorProphecy->getValue($dummy, 'children')->willReturn(null)->shouldBeCalled();

        $resourceClassResolverProphecy = $this->prophesize(ResourceClassResolverInterface::class);
        $resourceClassResolverProphecy->getResourceClass($dummy, DummyTableInheritanceRelated::class)->willReturn(DummyTableInheritanceRelated::class)->shouldBeCalled();

        $serializerProphecy = $this->prophesize(SerializerInterface::class);
        $serializerProphecy->willImplement(NormalizerInterface::class);
        $serializerProphecy->normalize(null, null, Argument::type('array'))->willReturn(null)->shouldBeCalled();

        $nullableNormalizer = $this->getMockForAbstractClass(AbstractItemNormalizer::class, [
            $propertyNameCollectionFactoryProphecy->reveal(),
            $nullablePropertyMetadataFactoryProphecy->reveal(),
            $iriConverterProphecy->reveal(),
            $resourceClassResolverProphecy->reveal(),
            $propertyAccessorProphecy->reveal(),
            null,
            null,
            null,
            false,
            [],
            [],
            null,
            false,
        ]);
        $nullableNormalizer->setSerializer($serializerProphecy->reveal());

        $notNullableNormalizer = $this->getMockForAbstractClass(AbstractItemNormalizer::class, [
            $propertyNameCollectionFactoryProphecy->reveal(),
            $notNullablePropertyMetadataFactoryProphecy->reveal(),
            $iriConverterProphecy->reveal(),
            $resourceClassResolverProphecy->reveal(),
            $propertyAccessorProphecy->reveal(),
            null,
            null,
            null,
            false,
            [],
            [],
            null,
            false,
        ]);
        $notNullableNormalizer->setSerializer($serializerProphecy->reveal());

        $this->assertEquals(
            ['children' => null],
            $nullableNormalizer->normalize($dummy, null, ['resource_class' => DummyTableInheritanceRelated::class, 'resources' => []])
        );

        $this->expectException(UnexpectedValueException::class);
        $notNullableNormalizer->normalize($dummy, null, ['resource_class' => DummyTableInheritanceRelated::class, 'resources' => []]);
    }

**Comment by teohhanhui** [comment]
I'm sorry, but `null` is not a valid value for a to-many relation. It makes no sense.

**Comment by teohhanhui** [comment]
> for my application, if a to-many relation is set to null, it's because I didn't hydrate that sub-resource.

We have no way of handling this in a general way.

**Comment by fherbin** [comment]
> I'm sorry, but `null` is not a valid value for a to-many relation. It makes no sense.

Okay, so if you have a solution to differentiate an empty hydrated to-many relation (an empty array) from a void for depth-limited sub-objects, I'm in.

**Pull request closed without merging**